sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala

```diff
@@ -132,14 +133,18 @@ object V2ScanRelationPushDown extends Rule[LogicalPlan] with PredicateHelper {
                         val count = aggregate.Count(avg.child).toAggregateExpression(isDistinct)
                         // Closely follow `Average.evaluateExpression`
                         avg.dataType match {
+                          case dt: DecimalType =>
+                            addCastIfNeeded(DecimalPrecision.decimalAndDecimal()(
+                              Divide(
+                                CheckOverflowInSum(sum, dt, !avg.failOnError),
```

Comment:
Since we need to push down SUM, we can only do it if ansi mode is true, otherwise the JDBC database may throw error for sum overflow and change the behavior of Spark AVG.